otri/validation/__init__.py

```diff
@@ -0,0 +1,572 @@
+                The checked data.
+
+            index : List[int]
+                The indexes from where each atom came from.
+        '''
+        for atom, index in zip(data, indexes):
+            self._push_data(atom, index)
+
+    def _on_result(self, data: List[Mapping], result: Any, indexes: List[int]):
+        '''
+        Called if `_check` returned something.
+        Labels all the atoms in the `data` list and pushes them all.
+
+        Parameters:
+            data : List[Mapping]
+                The checked data.
+
+            result : Any
+                The result of `_check`.
+
+            indexes : List[int]
+                The index of the input the data has been popped from.
+        '''
+        for atom, index in zip(data, indexes):
+            append_label(atom, result)
+            self._push_data(atom, index)
+
+    def _check(self, data: List[Mapping], indexes: List[int]):
+        '''
+        Parameters:
+            data : List[Mapping]
+                The atoms retrieved from the inputs.
+
+            indexes : List[int]
+                The indexes from which the atoms come from.
+
+        Raises:
+            NotImplementedError. This is an abstract class.
+        '''
+        return ValidatorFilter._check(self, data)
+
+    def _on_data(self, data: List[Mapping], indexes: List[int]):
+        '''
+        Called when input data is found.
+
+        Parameters:
+            data : List[Mapping]
+                The list of atoms from the inputs, one for each of the inputs that are still open.
+
+            indexes : List[int]
+                The indexes of the Streams from which the atoms come from.
+        '''
+        result = self._check(data, indexes)
+        if result is None:
+            self._on_ok(data, indexes)
+        else:
+            log.v(msg="{}. Data: {}\nAnalysis: {}.".format(self, data, result))
+            self._on_result(data, result, indexes)
+
+
+class ParallelBufferValidator(ParallelValidator):
+
+    '''
+    Class implementing a Filter with multiple buffers, reading from multiple Streams in parallel.
+    This is very similar to `BufferedValidator`, except the data is always passed as a list of
+    values coming from input Streams only when available.
+    There is only one internal buffer. `_buffer_top`, `_buffer_pop` and `_release` behave
+    accordingly.
+    '''
+
+    def __init__(self, inputs: Sequence[str], outputs: Sequence[str], check: Callable = None):
+        '''
+        This Validator expects the same number of Stream inputs and outputs.
+
+        Parameters:
+            inputs : str
+                Names of the inputs.
+
+            outputs : str
+                Names of the outputs.
+
+            check : Callable
+                If you don't want to override the class, you can pass a Callable here.
+                The Callable should require the atom batch as a parameter.
+        '''
+        ParallelFilter.__init__(self, inputs, outputs)
```

Comment:
Codacy found an issue: [__init__ method from a non direct base class 'ParallelFilter' is called](https://app.codacy.com/gh/OTRI-Unipd/OTRI/pullRequest?prid=6277039)